### DISCUSSION TITLE
Summarize and launch new task when context becomes too large

### DISCUSSION THREAD

**Discussion by napter** [cline/cline#747] [Feature Requests]
## Problem summary
One common issue when working on a project is that the context gets longer than what the model can really handle. It also gets particularly expensive. What I sometimes do manually is save the entire conversation, then send it to Gemeni since it has a 1M token context and ask it to provide the initial instructions, a summary a list of the action items, what has been so far and what is left to do - and to provide context and guidance for the continuation of the task.

If this could be a built in feature, I would use it often during a project to clean up the context and continue - sometimes it is absolutely necessary since the tokens exceed Claude's context window.

## Feature details
* Allow the user to select a model to use for summarizing and launching a new task. 
* Include a default prompt instructing the summarization model or let the user create one.
* Add the ability to "summarize and launch new task" to continue the task with reduced context

**Comment by halobartku** [comment]
Yeah, it ate me 3$ just like that popping exceed input token limit 6 times, eating 0.5$ each time. This is crazy. would be great to have it fixed.

**Comment by brentnhunter** [comment]
I agree!

**Comment by siscia** [comment]
I found that is better to do it manually to be honest. 

I found myself in the situation you guys are reporting, and I don't find the model able enough to figure out how t summarize what happened.

Sometimes, I just `git revert` and start from scratch, I get to the same status (or better) in just few prompts and I move from there.

**Comment by jamesrao** [comment]
In VSCode, when developing large-scale software projects using Cline, due to the maximum token output limit, only a portion of the work can be completed each time, and a new task must be restarted. However, Cline lacks a memory function, so restarting a new task means starting from the beginning again, resulting in repeated API requests, wasted tokens, and reduced development efficiency. Could the VSCode development team consider adding a memory function and increasing the token output limit for large projects?

**Reply by siscia** [reply]
But this is not how it works.

The token limit is imposed by the model. Claude is 200k, the latest Google one is 2M, and deepseek is 64k.

**Comment by FlorentPRABEL** [comment]
What I don't understand is, is it possible to pay more and not have this problem?

**Reply by joewilsonai** [reply]
No, the LLM models enforce the context limits. Cline could develop a context memory manager and optimizer to act as a "middleman," so when what you've been doing falls out of context, it can look back and bring back relevant parts of the context for the current task you are working on. But that could be a pretty complex problem to solve.

**Comment by atomraymaker** [comment]
I find that I only need to keep an eye on cline to prevent hitting too high context length. I use memory bank and ask to maintain a list of tasks. I often ask to summarize current state so I can start a new session. I can’t find no it now but I think it was cline on x recently saying how over 100k the models struggle and best to not let it get over 150k so I need to watch for that when otherwise it is working through the task fine. 

It would be nice to be able to configure a summarize and start new session context limit. 

Something like:
- Check context limit before sending
- if it is over the config then insert a configured “summarize” prompt instead of taking the next step
- start a new session and send the configured “continue” prompt. 
- maybe a config to track the max number of restarts. Or the existing api calls limit might catch that. Though it would be nice to take an action on hitting that also so it could notify elsewhere.

**Reply by atomraymaker** [reply]
Related to this. I was looking to hook cline up with slack or similar. Have it run a bags script and wait for exit when it has a question or blocker. The script could poll the chat for a response, echo, and exit. This should work fine today but without some check on the context limit it wouldn’t work for long. I am imagining running a half a dozen or more agents and not needing to check in unless there is a blocker.

**Comment by HavenCTO** [comment]
Are there any cline forks that support this feature?
